src/filetree.rs

```diff
@@ -302,34 +330,85 @@ pub(crate) fn apply_diff(
     }
     // Ensure all of the new files are written persistently to disk
     if !opts.skip_sync {
         syncfs(destdir)?;
     }
-    // Now move them all into place (TODO track interruption)
-    for path in diff.additions.iter().chain(diff.changes.iter()) {
-        let pathtmp = tmpname_for_path(path);
-        destdir
-            .local_rename(&pathtmp, path)
-            .with_context(|| format!("renaming {path}"))?;
+
+    // Write new files to temp dir if exists, else write directly in dest
+    for pathstr in diff.additions.iter() {
+        let path = Utf8Path::new(pathstr);
+        let (subdir, subdir_tmp) = get_subdir(path)?;
+        let path_tmp;
+        if destdir.exists(&subdir_tmp)? {
+            path_tmp = Utf8Path::new(&subdir_tmp).join(path.strip_prefix(&subdir)?);
+        } else {
+            path_tmp = path.to_path_buf();
+        }
+        // ensure new additions dir exists
+        if let Some(parent) = path_tmp.parent().filter(|&v| !v.as_os_str().is_empty()) {
+            destdir.ensure_dir_all(parent.as_std_path(), DEFAULT_FILE_MODE)?;
+        }
+        srcdir
+            .copy_file_at(path.as_std_path(), destdir, path_tmp.as_std_path())
+            .with_context(|| format!("copying {:?} to {:?}", path, path_tmp))?;
     }
-    if !opts.skip_removals {
-        for path in diff.removals.iter() {
+
+    // Write changed files to temp dir and rename finally
+    for pathstr in diff.changes.iter() {
+        let path = Utf8Path::new(pathstr);
+        let (subdir, subdir_tmp) = get_subdir(path)?;
+        let path_tmp;
+        if subdir != *pathstr {
+            if !destdir.exists(&subdir_tmp)? {
+                copy_dir(destdir, &subdir, &subdir_tmp)?;
+                updates.insert(subdir.clone(), subdir_tmp.clone());
+            }
+            path_tmp = Utf8Path::new(&subdir_tmp).join(path.strip_prefix(&subdir)?);
             destdir
-                .remove_file_optional(path)
-                .with_context(|| format!("removing {path}"))?;
+                .remove_file_optional(path_tmp.as_std_path())
+                .with_context(|| format!("removing {path_tmp}"))?;
+        } else {
+            // file that is not in dir, remember the changes to do rename
+            updates.insert(subdir.clone(), subdir_tmp.clone());
+            path_tmp = Utf8PathBuf::from(&subdir_tmp);
         }
+        srcdir
+            .copy_file_at(path.as_std_path(), destdir, path_tmp.as_std_path())
+            .with_context(|| format!("copying {:?} to {:?}", path, path_tmp))?;
+    }
+
+    // do local exchange
+    for (src, tmp) in updates {
+        log::trace!("doing local exchange for {} and {}", tmp, src);
+        destdir
+            .local_exchange(&tmp, &src)
+            .with_context(|| format!("exchange for {:?} and {:?}", tmp, src))?;
+        // finally remove the temp dir
+        log::trace!("cleanup: {:?}", tmp);
+        destdir.remove_all(&tmp).context("clean up temp")?;
```

Comment:
Ideally we do that in a second for loop to minimize the time between when we start exchanging the folders and when we complete it. The removal can append at the very end.